File: packages/sdks/client-sdks/gatrix-js-client-sdk/src/Metrics.ts
/**
 * Metrics - SDK usage tracking
 * Based on unleash-js-sdk metrics implementation
 *
 * Tracks:
 * - Flag access counts (yes/no)
 * - Variant usage counts
 * - Access to undefined flags
 */

import { Logger } from './Logger';
import { EventEmitter } from './EventEmitter';
import { EVENTS } from './events';
import ky from 'ky';

export interface MetricsOptions {
  appName: string;
  apiUrl: string;
  apiToken: string;
  environment: string;
  customHeaders?: Record<string, string>;
  metricsInterval?: number; // seconds (default: 60)
  metricsIntervalInitial?: number; // seconds before first send (default: 2)
  disableMetrics?: boolean;
  logger?: Logger;
  connectionId?: string;
  emitter?: EventEmitter;
}

interface VariantBucket {
  [variantName: string]: number;
}

interface FlagBucket {
  yes: number;
  no: number;
  variants: VariantBucket;
}

interface Bucket {
  start: Date;
  stop: Date | null;
  flags: {
    [flagName: string]: FlagBucket;
  };
  missing: {
    [flagName: string]: number;
  };
}

interface MetricsPayload {
  bucket: {
    start: Date;
    stop: Date | null;
    flags: { [flagName: string]: FlagBucket };
    missing: { [flagName: string]: number };
  };
  appName: string;
  instanceId: string;
}

export class Metrics {
  private bucket: Bucket;
  private appName: string;
  private apiUrl: string;
  private apiToken: string;
  private environment: string;
  private customHeaders: Record<string, string>;
  private metricsInterval: number;
  private metricsIntervalInitial: number;
  private disabled: boolean;
  private logger: Logger | undefined;
  private connectionId: string;
  private emitter: EventEmitter | undefined;
  private timer: ReturnType<typeof setInterval> | undefined;
  private started: boolean = false;

  constructor(options: MetricsOptions) {
    this.appName = options.appName;
    this.apiUrl = options.apiUrl;
    this.apiToken = options.apiToken;
    this.environment = options.environment;
    this.customHeaders = options.customHeaders ?? {};
    this.metricsInterval = (options.metricsInterval ?? 60) * 1000;
    this.metricsIntervalInitial = (options.metricsIntervalInitial ?? 2) * 1000;
    this.disabled = options.disableMetrics ?? false;
    this.logger = options.logger;
    this.connectionId = options.connectionId ?? '';
    this.emitter = options.emitter;
    this.bucket = this.createEmptyBucket();
  }

  /**
   * Start metrics collection and periodic sending
   */
  start(): void {
    if (this.disabled || this.started) {
      return;
    }
    this.started = true;

    if (this.metricsInterval > 0) {
      if (this.metricsIntervalInitial > 0) {
        setTimeout(() => {
          this.startTimer();
          this.sendMetrics();
        }, this.metricsIntervalInitial);
      } else {
        this.startTimer();
      }
    }
  }

  /**
   * Stop metrics collection
   */
  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = undefined;
    }
    this.started = false;
  }

  /**
   * Count flag access
   * @param flagName Flag name
   * @param enabled Whether the flag was enabled
   */
  count(flagName: string, enabled: boolean): void {
    if (this.disabled) return;

    this.assertBucket(flagName);
    this.bucket.flags[flagName][enabled ? 'yes' : 'no']++;
  }

  /**
   * Count variant usage
   * @param flagName Flag name
   * @param variantName Variant name
   */
  countVariant(flagName: string, variantName: string): void {
    if (this.disabled) return;

    this.assertBucket(flagName);
    const variants = this.bucket.flags[flagName].variants;
    variants[variantName] = (variants[variantName] ?? 0) + 1;
  }

  /**
   * Count access to missing (not found) flags
   * @param flagName Flag name that was not found
   */
  countMissing(flagName: string): void {
    if (this.disabled) return;

    this.bucket.missing[flagName] = (this.bucket.missing[flagName] ?? 0) + 1;
  }

  /**
   * Get missing flags record for statistics
   */
  getMissingFlags(): Record<string, number> {
    return { ...this.bucket.missing };
  }

  /**
   * Send current metrics to server
   */
  async sendMetrics(): Promise<void> {
    if (this.disabled) return;

    const payload = this.getPayload();
    if (this.bucketIsEmpty(payload)) {
      return;
    }

    try {
      const metricsUrl = this.buildMetricsUrl();
      await ky.post(metricsUrl, {
        headers: this.getHeaders(),
        json: payload,
        retry: {
          limit: 2,
          methods: ['post'],
          statusCodes: [408, 429, 500, 502, 503, 504],
        },
        timeout: 10000,
      });
      this.logger?.debug('Metrics sent successfully');
      this.emitter?.emit(EVENTS.METRICS_SENT, payload);
    } catch (e) {
      this.logger?.error('Failed to send metrics', e);
      this.emitter?.emit(EVENTS.METRICS_ERROR, e);
    }
  }

  private createEmptyBucket(): Bucket {
    return {
      start: new Date(),
      stop: null,
      flags: {},
      missing: {},
    };
  }

  private assertBucket(flagName: string): void {
    if (!this.bucket.flags[flagName]) {
      this.bucket.flags[flagName] = {
        yes: 0,
        no: 0,
        variants: {},
      };
    }
  }

  private startTimer(): void {
    this.timer = setInterval(() => {
      this.sendMetrics();
    }, this.metricsInterval);
  }

  private bucketIsEmpty(payload: MetricsPayload): boolean {
    return (
      Object.keys(payload.bucket.flags).length === 0 &&
      Object.keys(payload.bucket.missing).length === 0
    );
  }

  private getPayload(): MetricsPayload {
    const bucket = { ...this.bucket, stop: new Date() };
    this.bucket = this.createEmptyBucket();

    return {
      bucket,
      appName: this.appName,
      instanceId: this.connectionId,
    };
  }

  private buildMetricsUrl(): string {
    // Metrics endpoint: {apiUrl}/client/features/{environment}/metrics
    return `${this.apiUrl}/client/features/${this.environment}/metrics`;
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'X-API-Token': this.apiToken,
      'X-Application-Name': this.appName,
      'X-Connection-Id': this.connectionId,
      ...this.customHeaders,
    };
    return headers;
  }
}

File: packages/sdks/client-sdks/gatrix-unity-sdk/Dummy/TestBed/Assets/Gatrix/Runtime/Metrics.cs
// Metrics - SDK usage tracking for Gatrix Unity Client SDK
// Thread-safe: bucket access is locked, event emissions dispatched to main thread

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Gatrix.Unity.SDK
{
    /// <summary>
    /// Metrics payload structure for server submission
    /// </summary>
    internal class MetricsPayload
    {
        public string AppName { get; set; }
        public string InstanceId { get; set; }
        public MetricsBucket Bucket { get; set; }
    }

    /// <summary>
    /// Metrics bucket - tracks flag access counts
    /// </summary>
    internal class MetricsBucket
    {
        public DateTime Start { get; set; }
        public DateTime? Stop { get; set; }
        public Dictionary<string, FlagBucket> Flags { get; set; }
        public Dictionary<string, int> Missing { get; set; }
    }

    /// <summary>
    /// Per-flag metrics bucket
    /// </summary>
    internal class FlagBucket
    {
        public int Yes { get; set; }
        public int No { get; set; }
        public Dictionary<string, int> Variants { get; set; }

        public FlagBucket()
        {
            Variants = new Dictionary<string, int>();
        }
    }

    /// <summary>
    /// Metrics collection and periodic server submission.
    /// Thread-safe: bucket operations are locked since Count() is called
    /// from the main thread while SendMetrics runs on a background continuation.
    /// </summary>
    internal class Metrics
    {
        private readonly string _appName;
        private readonly string _apiUrl;
        private readonly string _apiToken;
        private readonly string _environment;
        private readonly Dictionary<string, string> _customHeaders;
        private readonly bool _disabled;
        private readonly IGatrixLogger _logger;
        private readonly string _connectionId;
        private readonly GatrixEventEmitter _emitter;
        private readonly HttpClient _httpClient;
        private readonly SynchronizationContext _syncContext;

        // Lock protects _bucket read/write from concurrent access
        private readonly object _bucketLock = new object();
        private MetricsBucket _bucket;
        private CancellationTokenSource _cts;
        private bool _started;

        public Metrics(
            string appName,
            string apiUrl,
            string apiToken,
            string environment,
            Dictionary<string, string> customHeaders,
            bool disableMetrics,
            IGatrixLogger logger,
            string connectionId,
            GatrixEventEmitter emitter,
            HttpClient httpClient)
        {
            _appName = appName;
            _apiUrl = apiUrl;
            _apiToken = apiToken;
            _environment = environment;
            _customHeaders = customHeaders;
            _disabled = disableMetrics;
            _logger = logger;
            _connectionId = connectionId ?? "";
            _emitter = emitter;
            _httpClient = httpClient;
            _bucket = CreateEmptyBucket();

            // Capture SynchronizationContext for main thread dispatch
            _syncContext = SynchronizationContext.Current;
        }

        /// <summary>Start metrics collection</summary>
        public void Start(int metricsIntervalMs = 60000, int metricsIntervalInitialMs = 2000)
        {
            if (_disabled)
            {
                LogOnMainThread("Metrics disabled, skipping start");
                return;
            }
            if (_started) return;
            _started = true;

            LogOnMainThread($"Metrics started. interval={metricsIntervalMs}ms, initialDelay={metricsIntervalInitialMs}ms");
            _cts = new CancellationTokenSource();
            _ = RunMetricsLoop(metricsIntervalMs, metricsIntervalInitialMs, _cts.Token);
        }

        /// <summary>Stop metrics collection</summary>
        public void Stop()
        {
            _cts?.Cancel();
            _cts?.Dispose();
            _cts = null;
            _started = false;
        }

        /// <summary>Count flag access (called from main thread)</summary>
        public void Count(string flagName, bool enabled)
        {
            if (_disabled) return;

            lock (_bucketLock)
            {
                var bucket = AssertBucket(flagName);
                if (enabled) bucket.Yes++;
                else bucket.No++;
            }
        }

        /// <summary>Count variant usage (called from main thread)</summary>
        public void CountVariant(string flagName, string variantName)
        {
            if (_disabled) return;

            lock (_bucketLock)
            {
                var bucket = AssertBucket(flagName);
                if (bucket.Variants.TryGetValue(variantName, out var count))
                {
                    bucket.Variants[variantName] = count + 1;
                }
                else
                {
                    bucket.Variants[variantName] = 1;
                }
            }
        }

        /// <summary>Count access to missing flags (called from main thread)</summary>
        public void CountMissing(string flagName)
        {
            if (_disabled) return;

            lock (_bucketLock)
            {
                if (_bucket.Missing.TryGetValue(flagName, out var count))
                {
                    _bucket.Missing[flagName] = count + 1;
                }
                else
                {
                    _bucket.Missing[flagName] = 1;
                }
            }
        }

        /// <summary>Get missing flags record for statistics</summary>
        public Dictionary<string, int> GetMissingFlags()
        {
            lock (_bucketLock)
            {
                return new Dictionary<string, int>(_bucket.Missing);
            }
        }

        /// <summary>Send current metrics to server</summary>
        public async ValueTask SendMetricsAsync()
        {
            if (_disabled) return;

            var payload = GetPayload();
            if (BucketIsEmpty(payload))
            {
                LogOnMainThread("Metrics: bucket empty, skipping send");
                return;
            }

            var flagCount = payload.Bucket.Flags.Count;
            var missingCount = payload.Bucket.Missing.Count;
            LogOnMainThread($"Metrics: sending. flags={flagCount}, missing={missingCount}");

            const int maxRetries = 2;
            var url = $"{_apiUrl}/client/features/{_environment}/metrics";
            var json = GatrixJson.SerializeMetrics(payload);

            for (int attempt = 0; attempt <= maxRetries; attempt++)
            {
                try
                {
                    var content = new StringContent(json, Encoding.UTF8, "application/json");
                    var request = new HttpRequestMessage(HttpMethod.Post, url)
                    {
                        Content = content
                    };
                    request.Headers.TryAddWithoutValidation("X-API-Token", _apiToken);
                    request.Headers.TryAddWithoutValidation("X-Application-Name", _appName);
                    request.Headers.TryAddWithoutValidation("X-Connection-Id", _connectionId);
                    request.Headers.TryAddWithoutValidation("X-SDK-Version", $"{GatrixClient.SdkName}/{GatrixClient.SdkVersion}");
                    if (_customHeaders != null)
                    {
                        foreach (var kvp in _customHeaders)
                        {
                            request.Headers.TryAddWithoutValidation(kvp.Key, kvp.Value);
                        }
                    }

                    var response = await _httpClient.SendAsync(request);
                    var statusCode = (int)response.StatusCode;

                    if (statusCode < 400)
                    {
                        _logger?.Debug("Metrics sent successfully");
                        EmitOnMainThread(GatrixEvents.FlagsMetricsSent, payload);
                        return;
                    }

                    // Retry on retryable status codes
                    bool retryable = statusCode == 408 || statusCode == 429 || statusCode >= 500;
                    if (retryable && attempt < maxRetries)
                    {
                        var delay = (int)Math.Pow(2, attempt + 1) * 1000;
                        await Task.Delay(delay);
                        continue;
                    }

                    var body = await response.Content.ReadAsStringAsync();
                    throw new Exception($"HTTP {statusCode}: {body}");
                }
                catch (Exception e)
                {
                    if (attempt >= maxRetries)
                    {
                        var errorMsg = $"Failed to send metrics after {maxRetries + 1} attempts: {e.Message}";
                        // Log on main thread to ensure visibility in Unity console
                        if (_syncContext != null)
                        {
                            _syncContext.Post(_ => _logger?.Error(errorMsg), null);
                        }
                        else
                        {
                            _logger?.Error(errorMsg);
                        }
                        EmitOnMainThread(GatrixEvents.FlagsMetricsError, e);
                    }
                    else
                    {
                        var delay = (int)Math.Pow(2, attempt + 1) * 1000;
                        await Task.Delay(delay);
                    }
                }
            }
        }

        // ==================== Private ====================

        private async ValueTask RunMetricsLoop(int intervalMs, int initialDelayMs, CancellationToken ct)
        {
            try
            {
                if (initialDelayMs > 0)
                {
                    await Task.Delay(initialDelayMs, ct);
                }

                while (!ct.IsCancellationRequested)
                {
                    await SendMetricsAsync();
                    await Task.Delay(intervalMs, ct);
                }
            }
            catch (OperationCanceledException)
            {
                // Expected on stop
            }
        }

        /// <summary>
        /// Emit event on the main thread to prevent Unity API calls from background threads
        /// </summary>
        private void EmitOnMainThread(string eventName, object data = null)
        {
            if (_syncContext != null && SynchronizationContext.Current != _syncContext)
            {
                _syncContext.Post(_ =>
                {
                    _emitter?.Emit(eventName, data);
                }, null);
            }
            else
            {
                _emitter?.Emit(eventName, data);
            }
        }

        /// <summary>
        /// Log debug message on the main thread to ensure visibility in Unity console
        /// </summary>
        private void LogOnMainThread(string message)
        {
            if (_syncContext != null && SynchronizationContext.Current != _syncContext)
            {
                _syncContext.Post(_ => _logger?.Debug(message), null);
            }
            else
            {
                _logger?.Debug(message);
            }
        }

        private MetricsBucket CreateEmptyBucket()
        {
            return new MetricsBucket
            {
                Start = DateTime.UtcNow,
                Stop = null,
                Flags = new Dictionary<string, FlagBucket>(),
                Missing = new Dictionary<string, int>()
            };
        }

        private FlagBucket AssertBucket(string flagName)
        {
            // Caller must hold _bucketLock
            if (!_bucket.Flags.TryGetValue(flagName, out var bucket))
            {
                bucket = new FlagBucket();
                _bucket.Flags[flagName] = bucket;
            }
            return bucket;
        }

        private bool BucketIsEmpty(MetricsPayload payload)
        {
            return payload.Bucket.Flags.Count == 0 && payload.Bucket.Missing.Count == 0;
        }

        private MetricsPayload GetPayload()
        {
            lock (_bucketLock)
            {
                var bucket = _bucket;
                bucket.Stop = DateTime.UtcNow;
                _bucket = CreateEmptyBucket();

                return new MetricsPayload
                {
                    Bucket = bucket,
                    AppName = _appName,
                    InstanceId = _connectionId
                };
            }
        }
    }
}
